// redux-state/lib/staty.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _actionTypes = require('./actionTypes');

var _connectState = require('./connectState');

var _connectState2 = _interopRequireDefault(_connectState);

var _statyReducer = require('./statyReducer');

var _statyReducer2 = _interopRequireDefault(_statyReducer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var mapStateToProps = function mapStateToProps(state) {
    return {
        state: state
    };
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
    return {
        setState: function setState(state) {
            return dispatch({
                type: _actionTypes.UPDATE_STATY,
                payload: {
                    state: state
                }
            });
        }
    };
};

var staty = function staty(initialState) {
    return function (WrappedComponent) {
        return (0, _connectState2["default"])(mapStateToProps, mapDispatchToProps, undefined, (0, _statyReducer2["default"])(initialState))(WrappedComponent);
    };
};

exports["default"] = staty;
